Source/Core/Core/IOS/ES/NandUtils.cpp

```diff
@@ -200,65 +200,96 @@ std::vector<std::array<u8, 20>> ES::GetSharedContents() const
+                         FS::Mode::ReadWrite, FS::Mode::None) != FS::ResultCode::Success)
+  {
+    ERROR_LOG(IOS_ES, "InitImport: Failed to create content dir for %016" PRIx64, tmd.GetTitleId());
+    return false;
   }
 
-  IOS::ES::UIDSys uid_sys{m_ios.GetFS()};
-  uid_sys.GetOrInsertUIDForTitle(title_id);
+  const std::string data_dir = Common::GetTitleDataPath(tmd.GetTitleId());
+  const auto data_dir_contents = fs->ReadDirectory(0, 0, data_dir);
+  if (!data_dir_contents &&
+      (data_dir_contents.Error() != FS::ResultCode::NotFound ||
+       fs->CreateDirectory(0, 0, data_dir, 0, FS::Mode::ReadWrite, FS::Mode::None,
+                           FS::Mode::None) != FS::ResultCode::Success))
+  {
+    return false;
+  }
 
-  // IOS moves the title content directory to /import if the TMD exists during an import.
-  if (File::Exists(Common::GetTMDFileName(title_id, Common::FROM_SESSION_ROOT)))
+  IOS::ES::UIDSys uid_sys{fs};
+  const u32 uid = uid_sys.GetOrInsertUIDForTitle(tmd.GetTitleId());
+  if (fs->SetMetadata(0, data_dir, uid, tmd.GetGroupId(), 0, FS::Mode::ReadWrite, FS::Mode::None,
+                      FS::Mode::None) != FS::ResultCode::Success)
   {
-    const std::string import_content_dir =
-        Common::GetImportTitlePath(title_id, Common::FROM_SESSION_ROOT) + "/content";
-    File::CreateFullPath(import_content_dir);
-    if (!File::Rename(content_dir, import_content_dir))
-    {
-      ERROR_LOG(IOS_ES, "InitImport: Failed to move content dir for %016" PRIx64, title_id);
-      return false;
-    }
+    return false;
   }
 
+  // IOS moves the title content directory to /import if the TMD exists during an import.
+  const auto file_info = fs->GetMetadata(0, 0, Common::GetTMDFileName(tmd.GetTitleId()));
+  if (!file_info || !file_info->is_file)
+    return true;
+
+  const auto result = fs->Rename(0, 0, content_dir, import_content_dir);
+  if (result != FS::ResultCode::Success)
+  {
+    ERROR_LOG(IOS_ES, "InitImport: Failed to move content dir for %016" PRIx64, tmd.GetTitleId());
+    return false;
+  }
+  DeleteDirectoriesIfEmpty(m_ios.GetFS().get(), import_content_dir);
   return true;
 }
 
 bool ES::FinishImport(const IOS::ES::TMDReader& tmd)
 {
   const u64 title_id = tmd.GetTitleId();
-  const std::string import_content_dir =
-      Common::GetImportTitlePath(title_id, Common::FROM_SESSION_ROOT) + "/content";
+  const std::string import_content_dir = Common::GetImportTitlePath(title_id) + "/content";
 
   // Remove everything not listed in the TMD.
   std::unordered_set<std::string> expected_entries = {"title.tmd"};
   for (const auto& content_info : tmd.GetContents())
     expected_entries.insert(StringFromFormat("%08x.app", content_info.id));
-  const auto entries = File::ScanDirectoryTree(import_content_dir, false);
-  for (const File::FSTEntry& entry : entries.children)
+  const auto entries = m_ios.GetFS()->ReadDirectory(0, 0, import_content_dir);
```

Comment:
Thoughts on storing the FS result for later instead of re-getting it all the time? (`auto fs = m_ios.GetFS()` like in all the other places)

Reply:
Done.